Create sessions

fordim opened the pull request:
ТЗ до 1.07.2019
1.Сделать кнопку Logout
2.Загрузку всех страниц осуществить через index.php (лишнее удалить)
3.Посылаем POST запросы в index.php и обрабатываем их при помощи функций с function.php (Выводим обработанные данные на новой странице)
4.Сделать авторизацию через SESSION, при авторизации давать доступ к страницам cabinet, history. Без авторизации signIn, signUp. Страница main доступна всегда
5.При авторизации делать проверку с admin пользователем (пароль проверять через Hash)
6.Сделать logout из сессии

fordim committed:
create branch

fordim committed:
create session, logout, and put all pages in index.php

olegsklyarov commented on inc/functions.php:
```diff
@@ -20,15 +19,35 @@ function renderTemplate(string $name, array $data = []): string
+    print("<br>PasswordHash: " . $passwordHash);
 }
 
 function processFormSignIn(string $email, string $password){
```
Функция проверки логина и пароля пользователя должна сделать следующее:
1. Запросить в таблице User запись (строку таблицы) пользователя по значению аргумента `$email` (в нашем случае БД пока нет, поэтому просто проверяем, что `$email === $adminLogin`). Если `$email` не найден в БД, то функция возвращает `null`.
2. Проверить, что пароль верный при помощи функции `password_verify`. Если пароль неверный, то возвращаем `null`.
3. В этот шаг попадаем только если пользователь найден в БД и пароль верный. Тогда функция возвращает все данные пользователя (строка из таблицы User, соответствующая пользователю). Это будем ассоциативный массив с ключами из колонок таблицы: id, email, password, created_at... Напрмер:
```
[
  'id' => 1242,
  'name' => 'Дмитрий',
  'password_hash' => '$2y$10$DRgiLAg6mB1xFGmnVZqOm.0rd.He75/aEdTjXqVTpiV7SkvSt2OWG', // password = dummy
  'created_at' => '2019-06-03',
  'updated_at' => '2019-06-03',
]
```

Давай переименуем функцию в 
```
function findUserDataByEmailAndPassword(string $email, string $password): ?array { ... }
````

olegsklyarov commented on inc/functions.php:
```diff
@@ -20,15 +19,35 @@ function renderTemplate(string $name, array $data = []): string
 
 function processFormSignUp(string $name, string $email, string $password){
-
+    $passwordHash = password_hash($password, PASSWORD_DEFAULT);
```
Функция должна создавать в БД нового пользователя по алгоритму:
1. Запускаем `findUserDataByEmailAndPassword`, если функция вернула **не** `null`, то кидаем исключение с ошибкой. Это будет означать, что пользователь с таким `$email` уже зарегистрирован, второй раз регистрироваться нельзя. 
2. Проверяем $email при помощи функции [filter_val()](https://www.php.net/manual/ru/function.filter-var.php) с флагом `FILTER_VALIDATE_EMAIL`. Если значение `$email` не валидно, то кидаем исключение c ошибкой.
3. Запускаем запрос в БД на создание новой записи в таблицу `User`. Пароль в БД храним, вместо этого записываем `password_hash`. Если SQL операция выполнена успешно, функция возвращаем id созданной записи в БД [mysqli_insert_id()](https://www.php.net/manual/ru/mysqli.insert-id.php). В противном случае кидаем исключение с ошибкой от СУБД.

Переименовываем функцию в
```
function createUser(string $email, string $password, string $name): int { ... }
```

Исключение бросаем при помощи `throw`. Например:
```
throw new Exception("User with email $email already exists");
```

Ловить исключение нужно в контроллере (index.php) при помощи try - catch. Например:
```
try {
  createUser(...);
} catch ($e) {
  die($e->getMessage());
}
```

olegsklyarov commented on index.php:
```diff
@@ -7,32 +7,96 @@
 
 require_once('inc/functions.php');
 
+if (isset($_POST)) {
```
Условие всегда выполняется, убираем его.

olegsklyarov commented on index.php:
```diff
@@ -7,32 +7,96 @@
+                    'jsStyle' => 'js/history.js',
+                    'content' => renderTemplate('itemHistory.php'),
+                    'scriptForDate' =>
+                        '<script type="text/javascript" src="https://cdn.jsdelivr.net/jquery/latest/jquery.min.js"></script>
```
В контроллере (MVC) не должно быть HTML кода. Убираем этот код во View (MVC). Отсюда можно передать какой-либо флаг, чтобы в template знал какой HTML генерировать. Либо завести разные мини-шаблоны для разных `<script ...`

olegsklyarov reviewed (commented):
Проверил частично. Но текущих комментариев достаточно для работы. На следующей итерации возможно добавлю новых замечаний. В целом — гуд 👍

olegsklyarov committed:
* Show history data from getUserExpenses (remove hardcode from itemHistory.php)
                       * Setup error reporting

fordim committed:
Fixed problem with $_POST in index.php
Created => Adding a new record to the table history from UI
Created => Adding a new record to the table user from UI
Showed history from DB accounting

fordim committed:
Changes that was made during the call with Oleg 23.08
Created function fetchData
Optimized function getUserExpenses
Put all connection with MySQL to init.php

fordim committed:
Merge branch 'master' into create-sessions
